gnome-initial-setup/eos-demo-mode

```diff
@@ -0,0 +1,118 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+import argparse
+import glob
+import json
+import logging
+import os
+import subprocess
+import sys
+
+from gi.repository import GLib, Gio
+
+log = logging.getLogger(sys.argv[0])
+
+ICON_GRID_DIR = '/var/lib/eos-image-defaults/icon-grid'
+
+GOOGLE_CHROME = 'google-chrome.desktop'
+CHROMIUM_BROWSER = 'chromium-browser.desktop'
+
+GDM_DAEMON_CONF = '/etc/gdm3/daemon.conf'
+
+
+def dq(s):
+    return '"{}"'.format(s)
+
+
+def remove_chrome_from_icon_grids():
+    log.info("Restoring Chromium to icon grids, if needed")
+
+    pattern = os.path.join(ICON_GRID_DIR, 'icon-grid-*.json')
+    for path in glob.glob(pattern):
+        log.info("Checking %s", path)
+        try:
+            with open(path, 'r') as f:
+                grid = f.read()
+
+            if GOOGLE_CHROME in grid:
+                log.info("Updating %s", path)
+                grid = grid.replace(dq(GOOGLE_CHROME), dq(CHROMIUM_BROWSER))
+
+                with open(path, 'w') as f:
+                    f.write(grid)
+        except:
+            log.exception("while processing %s", path)
+
+
+def force_automatic_login_for_gdm(user):
+    keyfile = GLib.KeyFile.new()
+    keyfile.load_from_file(GDM_DAEMON_CONF, GLib.KeyFileFlags.NONE)
+    keyfile.set_boolean('daemon', 'AutomaticLoginEnable', True)
+    keyfile.set_string('daemon', 'AutomaticLogin', user)
+    keyfile.set_boolean('daemon', 'ForceAutomaticLogin', True)
+    keyfile.save_to_file(GDM_DAEMON_CONF)
+
+    # Now get a GDBus connection to systemd and reload the
+    # gdm unit so that our changes don't require a reboot
+    # to apply
+    bus = Gio.bus_get_sync(Gio.BusType.SYSTEM, None)
+    bus.call_sync('org.freedesktop.systemd1',
+                  '/org/freedesktop/systemd1',
+                  'org.freedesktop.systemd1.Manager',
+                  'ReloadUnit',
+                  GLib.Variant('(ss)', ('gdm.service', 'replace')),
+                  None,
+                  Gio.DBusCallFlags.NONE,
+                  -1,
+                  None)
+
+
+def enter_test_mode():
+    """Enter test mode.
+    
+    All filesystem changes past this point are now ephemeral.
+    """
+    test_mode_script = os.path.join(os.path.dirname(__file__), "eos-test-mode")
+    subprocess.check_call([test_mode_script])
+
+
+def touch_stamp_file(username):
+    """Touch a stamp file indicating that we're in demo mode."""
+    with open("/run/gnome-initial-setup/eos-demo-mode", "w") as f:
```

Comment:
No, it is different. We need to have a file in `/run/gnome-initial-setup` here so that gdm knows which users are "demo mode" users (which is different to the file touched in `setup_demo_config()` which is touched in the homedir, so that the autostart logic knows to launch `eos-demo-video` at the beginning of the user's session.